### DISCUSSION TITLE
Convergence errors with `zoobukhov8` in `canopy_air_coms.f90`

### DISCUSSION THREAD

**Discussion by Aariq** [EDmodel/ED2#380] [Q&A] [answered]
I've been getting errors in some, but not all, ED2 runs that are part of an ensemble run (using `PEcAn`).  Many of them are similar to the one below:

```
--------------------------------------------------------------
                     !!! FATAL ERROR !!!                      
--------------------------------------------------------------
    ---> File:        canopy_air_coms.f90
    ---> Subroutine:  zoobukhov8
    ---> Reason:      Zeta didn't converge, giving up!!!
--------------------------------------------------------------
```

Sometimes the reason is "Failed finding the second guess for regula falsi".

Are there any "usual suspects"—e.g. values in config.xml—that I should check to figure out why this is happening?

**Comment by larsonej** [comment]
This can happen if under certain environmental conditions. Is the wind very strong? I'd check the met drivers during the time they are crashing. Also check that your reference height (zref) is larger than your vegetation? This is required if you're using icanturb = 2. You can also try a different icanturb.
________________________________
From: Eric R. Scott ***@***.***>
Sent: Wednesday, February 15, 2023 2:55 PM
To: EDmodel/ED2 ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [EDmodel/ED2] Convergence errors with `zoobukhov8` in `canopy_air_coms.f90` (Discussion #380)


I've been getting errors in some, but not all, ED2 runs that are part of an ensemble run (using PEcAn). Many of them are similar to the one below:

--------------------------------------------------------------

                     !!! FATAL ERROR !!!

--------------------------------------------------------------

    ---> File:        canopy_air_coms.f90

    ---> Subroutine:  zoobukhov8

    ---> Reason:      Zeta didn't converge, giving up!!!

--------------------------------------------------------------



Sometimes the reason is "Failed finding the second guess for regula falsi".

Are there any "usual suspects"—e.g. values in config.xml—that I should check to figure out why this is happening?

**Reply by Aariq** [reply]
is zref something set in ED2IN?

**Reply by Aariq** [reply]
Found this issue: https://github.com/EDmodel/ED2/issues/58#issuecomment-94295732, which looks like what you're talking about.

My `zstar` is indeed less than `roughness` when I get this error. That issue identifies this as an ED2 bug (a sanity check gets skipped if b1Ht is in config.xml), and it's not clear if it got fixed.  I'll try removing b1HT from the config.xml and see if that works.

**Comment by mpaiao** [comment]
I had this issue in the past due to some variable not being correctly initialised in the cohort or patch dynamics step. If the crash is happening on the first day of the month, then it is likely not a problem with zoobukhov8, it is just that the function received a NaN that caused the solution to fail.

**Reply by Aariq** [reply]
This seems like a more likely culprit.  In a more recent attempt at an ensemble run, I'm getting this along with budget check errors, "attempt to set a site with 0 patches" errors, and some SIGABRT errors.

**Comment by Aariq** [comment, marked as answer]
I made my grasses and forbs have a higher DBH (1 cm instead of 0.6) in my .css file and it seems to have solved all the bugs (except the SIGABRT on the last day of simulation).  Can't remember where I saw it, but there was an issue in the repo that made me think that small and/or sparse plants were causing ED2 to have difficulty calculating air turbulence or something.  I have no clue what a DBH of 1cm means for a grass, but hopefully it is a reasonable fix??
